File: testdata/java_sources/StringUtils.java
package com.example;

/**
 * String utility functions.
 */
public class StringUtils {

    private StringUtils() {}

    /**
     * Reverse a string.
     */
    public static String reverse(String s) {
        return new StringBuilder(s).reverse().toString();
    }

    /**
     * Check if a string is a palindrome.
     */
    public static boolean isPalindrome(String s) {
        String cleaned = s.replaceAll("[^a-zA-Z0-9]", "").toLowerCase();
        return cleaned.equals(reverse(cleaned));
    }

    /**
     * Truncate a string to max length - dead code.
     */
    public static String truncate(String s, int max) {
        if (s.length() <= max) return s;
        return s.substring(0, max) + "...";
    }

    /**
     * Pad a string on the left - dead code.
     */
    public static String padLeft(String s, int width, char pad) {
        if (s.length() >= width) return s;
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < width - s.length(); i++) {
            sb.append(pad);
        }
        return sb.append(s).toString();
    }

    /**
     * Capitalize first letter - dead code.
     */
    public static String capitalize(String s) {
        if (s == null || s.isEmpty()) return s;
        return Character.toUpperCase(s.charAt(0)) + s.substring(1).toLowerCase();
    }
}

File: testdata/rust_sources/string_utils.rs
/// String utilities

/// Reverse a string
pub fn reverse(s: &str) -> String {
    s.chars().rev().collect()
}

/// Count occurrences of a char in a string
pub fn count_char(s: &str, c: char) -> usize {
    s.chars().filter(|&ch| ch == c).count()
}

/// Trim and lowercase a string
pub fn normalize(s: &str) -> String {
    s.trim().to_lowercase()
}

/// Check if string is a palindrome - dead code
pub fn is_palindrome(s: &str) -> bool {
    let normalized: String = s.chars().filter(|c| c.is_alphanumeric()).collect();
    let normalized = normalized.to_lowercase();
    normalized == reverse(&normalized)
}

/// Join strings with separator - dead code
pub fn join(parts: &[&str], sep: &str) -> String {
    parts.join(sep)
}

/// Split and trim - dead code
fn split_trim(s: &str, sep: char) -> Vec<String> {
    s.split(sep).map(|p| p.trim().to_string()).collect()
}

/// Capitalize first letter - dead code
pub fn capitalize(s: &str) -> String {
    let mut c = s.chars();
    match c.next() {
        None => String::new(),
        Some(f) => f.to_uppercase().collect::<String>() + c.as_str(),
    }
}
